Mathlib/RepresentationTheory/Coinvariants.lean

```diff
@@ -0,0 +1,182 @@
+def Coinvariants := V ⧸ augmentationSubmodule ρ
+
+namespace Coinvariants
+
+instance : AddCommGroup (Coinvariants ρ) := by unfold Coinvariants; infer_instance
+instance : Module k (Coinvariants ρ) := by unfold Coinvariants; infer_instance
+
+/-- The quotient map from a representation to its coinvariants as a linear map. -/
+def mk : V →ₗ[k] Coinvariants ρ := Submodule.mkQ (augmentationSubmodule ρ)
+
+theorem mk_eq_iff {x y : V} :
+    mk ρ x = mk ρ y ↔ x - y ∈ ρ.augmentationSubmodule :=
+  Submodule.Quotient.eq _
+
+@[simp]
+lemma mk_ρ_apply (g : G) (x : V) :
+    mk ρ (ρ g x) = mk ρ x :=
+  (mk_eq_iff _).2 <| mem_augmentationSubmodule_of_eq g x _ rfl
+
+@[elab_as_elim]
+theorem induction_on {C : Coinvariants ρ → Prop} (x : Coinvariants ρ) (h : ∀ v : V, C (mk ρ v)) :
+    C x :=
+  Submodule.Quotient.induction_on _ x h
+
+/-- A `G`-invariant linear map induces a linear map out of the coinvariants of a
+`G`-representation. -/
+def lift (f : V →ₗ[k] W) (h : ∀ (x : G), f ∘ₗ ρ x = f) :
+    ρ.Coinvariants →ₗ[k] W :=
+  Submodule.liftQ _ f <| Submodule.span_le.2 fun x ⟨⟨g, y⟩, hy⟩ => by
+    simpa only [← hy, SetLike.mem_coe, LinearMap.mem_ker, map_sub, sub_eq_zero, LinearMap.coe_comp,
+      Function.comp_apply] using LinearMap.ext_iff.1 (h g) y
+
+variable {ρ}
+
+@[simp]
+theorem lift_comp_mk (f : V →ₗ[k] W) (h : ∀ (x : G), f ∘ₗ ρ x = f) :
+    lift ρ f h ∘ₗ mk ρ = f := rfl
+
+@[simp]
+theorem lift_mk (f : V →ₗ[k] W) (h : ∀ (x : G), f ∘ₗ ρ x = f) (x : V) :
+    lift ρ f h (mk _ x) = f x := rfl
+
+@[ext]
+lemma hom_ext {f g : Coinvariants ρ →ₗ[k] W} (H : f ∘ₗ mk ρ = g ∘ₗ mk ρ) : f = g :=
+  Submodule.linearMap_qext _ H
+
+variable (ρ)
+
+/-- Given `G`-representations on `k`-modules `V, W`, a linear map `V →ₗ[k] W` commuting with
+the representations induces a `k`-linear map between the coinvariants. -/
+noncomputable def map (f : V →ₗ[k] W) (hf : ∀ g, f ∘ₗ ρ g = τ g ∘ₗ f) :
+    Coinvariants ρ →ₗ[k] Coinvariants τ :=
+  lift _ (mk _ ∘ₗ f) fun g => LinearMap.ext fun x => (mk_eq_iff _).2 <|
+    mem_augmentationSubmodule_of_eq (ρ := τ) g (f x) _ <| by simpa using congr($((hf g).symm) x)
+
+variable {ρ τ}
+
+@[simp]
+lemma map_comp_mk {f : V →ₗ[k] W} {hf : ∀ g, f ∘ₗ ρ g = τ g ∘ₗ f} :
+    map ρ τ f hf ∘ₗ mk ρ = mk τ ∘ₗ f := rfl
+
+@[simp]
+lemma map_mk (f : V →ₗ[k] W) (hf : ∀ g, f ∘ₗ ρ g = τ g ∘ₗ f) (x : V) :
+    map ρ τ f hf (mk _ x) = mk _ (f x) := rfl
+
+end Coinvariants
+end Representation
+
+namespace Rep
+
+open CategoryTheory Representation
+
+namespace Coinvariants
+
+variable {k G : Type u} [CommRing k] [Monoid G] {A B C : Rep k G} {n : ℕ}
+
+/-- The linear map underlying a `G`-representation morphism `A ⟶ B`, where `B` has the trivial
+representation, factors through `A_G`. -/
+noncomputable abbrev lift [B.ρ.IsTrivial] (f : A ⟶ B) :
+    A.ρ.Coinvariants →ₗ[k] B :=
+  Representation.Coinvariants.lift _ f.hom.hom fun _ => by
+    ext
+    have := hom_comm_apply f
+    simp_all
```

Comment:
How about this is replaced by `desc` and the output is a morphism in the category?